### PR TITLE
fix(consensus): handle Elapsed timeout at AwaitUtxo call site

### LINKED ISSUE #10629: fix(consensus): InternalDowncastError(Elapsed) near-tip sync stall after mempool bypass removal
## Summary

Zebra nodes near the tip are hitting repeated `InternalDowncastError(... Elapsed(()))` errors, causing a sync stall loop despite having 60+ peers.

## Root cause

The `From<BoxError> for TransactionError` impl in `zebra-consensus/src/error.rs` does not try downcasting to `tower::timeout::error::Elapsed`. When a timeout fires on:

- **`Timeout<State>` AwaitUtxo** (6-minute `UTXO_LOOKUP_TIMEOUT`) — line ~722 in `transaction.rs`, where `query.await?` converts the error via `From<BoxError>` with no `Elapsed` handling
- **`Timeout<Mempool>` AwaitOutput** (60s `MEMPOOL_OUTPUT_LOOKUP_TIMEOUT`) — line ~746, which attempts `err.downcast::<Elapsed>()` but falls back to `Err(err.into())` on failure, hitting the same unhandled path

...the `Elapsed` error falls through all downcast attempts (RedJubjub, ValidateContextError, TransactionError) and lands on `InternalDowncastError`.

The primary path is the **state AwaitUtxo** call at line ~722. For non-mempool block transactions, the `?` operator converts `Elapsed` directly through `From<BoxError> for TransactionError`, which doesn't know about `Elapsed` at all. This is the path for block verification at ~1,419 blocks from the tip, where the mempool may not even be active yet.

## Why this surfaced now

This bug has been latent since #8857 (Nov 2024) which introduced the `Timeout<Mempool>` wrapping and the `AwaitOutput` path.

**#10494 (Apr 17, 2026) removed the `find_verified_unmined_tx` shortcut** ("consensus: remove mempool bypass") for security reasons. Before that removal:

- The shortcut would find mempool transactions early and return cached verification results, skipping full verification entirely
- Timeout errors from the shortcut's mempool lookup were silently swallowed via `.ok()?`
- Most mempool transactions never reached the `AwaitOutput` path

After the removal, ALL transactions go through full verification, so transactions spending mempool outputs consistently hit the `AwaitOutput` path, and state UTXO lookups that time out now produce unhandled `Elapsed` errors.

## Fix

Add an `Elapsed` downcast arm to `From<BoxError> for TransactionError` in `zebra-consensus/src/error.rs`, mapping it to `TransactionError::TransparentInputNotFound` (consistent with how the explicit `downcast::<Elapsed>()` at line ~746 handles it).

## Affected code

**Primary path** (`zebra-consensus/src/transaction.rs`, line ~722):
```rust
// query.await? converts Elapsed through From<BoxError> for TransactionError
let query = state.clone().oneshot(zebra_state::Request::AwaitUtxo(*outpoint));
if let zebra_state::Response::Utxo(utxo) = query.await? {
```

**Secondary path** (`zebra-consensus/src/transaction.rs`, line ~746):
```rust
Err(err) => {
    return match err.downcast::<Elapsed>() {
        Ok(_) => Err(TransactionError::TransparentInputNotFound),
        Err(err) => Err(err.into()), // falls through to InternalDowncastError
    };
}
```

**Missing handler** (`zebra-consensus/src/error.rs`, line ~298):
```rust
impl From<BoxError> for TransactionError {
    fn from(mut err: BoxError) -> Self {
        // tries RedJubjub, ValidateContextError, TransactionError
        // does NOT try Elapsed — falls through to InternalDowncastError
    }
}
```

## Related

- #8857 — introduced the `AwaitOutput` + `Timeout<Mempool>` path (Nov 2024)
- #10494 — removed the `find_verified_unmined_tx` shortcut that masked the bug (Apr 2026)
- Not the same as #10627 (RouterError downcast mismatch in inbound gossip — different code path)

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Summary

- Handle `Elapsed` timeout at the `AwaitUtxo` call site in the transaction verifier, mapping it to `TransparentInputNotFound` instead of falling through to `InternalDowncastError`
- Mirrors the existing `AwaitOutput` pattern at line ~746
- Add unit test proving the fix
- Add CHANGELOG entry

## Motivation

Nodes near the tip hit repeated `InternalDowncastError(... Elapsed(()))` errors, causing a sync stall loop despite having 60+ peers.

The `From<BoxError> for TransactionError` impl tries downcasting to `RedJubjub`, `ValidateContextError`, and `TransactionError`, but not `tower::timeout::error::Elapsed`. When `Timeout<State>` (AwaitUtxo, 6 min) fires, the `Elapsed` error falls through to `InternalDowncastError`, failing block verification and triggering a sync restart loop.

This was latent since #8857 (Nov 2024) but became user-visible after #10494 (Apr 2026) removed the `find_verified_unmined_tx` shortcut that previously avoided or silently swallowed these timeouts.

## Solution

Handle `Elapsed` at the specific `AwaitUtxo` call site rather than in the broad `From<BoxError>` conversion. This avoids mislabelling unrelated state timeouts (e.g. `CheckBestChainTipNullifiersAndAnchors`) as `TransparentInputNotFound`.

The two timeout handling sites now use identical patterns:

```rust
Err(err) => return match err.downcast::<Elapsed>() {
    Ok(_) => Err(TransactionError::TransparentInputNotFound),
    Err(err) => Err(err.into()),
},
```

## Test plan

- [x] New unit test `block_request_with_await_utxo_timeout_returns_transparent_input_not_found` verifies that a mock `AwaitUtxo` returning `Elapsed` produces `TransparentInputNotFound`
- [x] Existing transaction verifier tests pass

Closes #10629

### CODE REVIEW

**Comment by oxarbitrage** [comment]
## Analysis: how the fix works

### Before the fix

When a block is verified near the tip, the transaction verifier looks up UTXOs from the state service, which is wrapped in `Timeout<ZS>` (6-minute `UTXO_LOOKUP_TIMEOUT`). If the state is slow and the timeout fires, Tower produces a `BoxError` containing `Elapsed`. The `?` operator on line ~722 converts this via `From<BoxError> for TransactionError`:

```
query.await?
  → From<BoxError> for TransactionError
  → Tries downcast::<RedJubjub>() — no
  → Tries downcast::<ValidateContextError>() — no
  → Tries downcast::<TransactionError>() — no
  → Nothing left to try
  → InternalDowncastError("...Elapsed(())")
```

`InternalDowncastError` is a catch-all for errors the verifier can't classify. The verifier DID understand what happened (a timeout), it just couldn't express it.

### After the fix

```
  → Tries downcast::<RedJubjub>() — no
  → Tries downcast::<ValidateContextError>() — no
  → Tries downcast::<TransactionError>() — no
  → Tries downcast::<Elapsed>() — YES
  → Returns TransparentInputNotFound
```

### Why TransparentInputNotFound is the correct error

That's exactly what happened — the transparent input's UTXO wasn't found within the timeout window. This is the same error that the explicit `downcast::<Elapsed>()` on line ~746 (the mempool `AwaitOutput` path) was already mapping to.

### Does this fully fix the stall loop?

This fix makes the error classification **correct**, which ensures the retry machinery works as designed. `TransparentInputNotFound` is a known, expected transaction error — the sync loop will retry the block, and on retry the state should have caught up. With `InternalDowncastError`, the error is unrecognized and could be scored or handled differently in ways that make retries less effective.

However, the deeper question is why the state service is timing out in the first place (6 minutes is a long timeout). If the state is persistently slow, retries will keep timing out regardless. That would be a separate issue worth investigating if the stall loop persists after this fix.

**Review by conradoplg — CHANGES_REQUESTED** [review]
Not sure about this - handling any timeout as a specific error seems error-prone. (I also don't think it would fix the sync issue, but might help debugging it)

Here is Claude's take - I have its suggested fix ready if you want me to commit it 

```
● You're right to be suspicious — the fix works for the immediate bug but it's
  overly broad.

  What flows through From<BoxError> for TransactionError

  The state field is Timeout<ZS> with UTXO_LOOKUP_TIMEOUT (6 min) and mempool is
   Timeout<Mempool> with MEMPOOL_OUTPUT_LOOKUP_TIMEOUT (60 s). Inside
  Verifier::call, four state/mempool queries are issued. Their error handling
  differs:

  transaction.rs:line: 703
  Query: state.oneshot(UnspentBestChainUtxo)
  How its error reaches TransactionError: explicit map_err(|_| 
    TransparentInputNotFound)
  ────────────────────────────────────────
  transaction.rs:line: 721
  Query: state.oneshot(AwaitUtxo)
  How its error reaches TransactionError: bare ? → goes through From<BoxError>
  ────────────────────────────────────────
  transaction.rs:line: 524
  Query: state.oneshot(CheckBestChainTipNullifiersAndAnchors)
  How its error reaches TransactionError: bundled into async_checks, surfaced
  via
    ? at 541 → goes through From<BoxError>
  ────────────────────────────────────────
  transaction.rs:line: 740
  Query: mempool.oneshot(AwaitOutput)
  How its error reaches TransactionError: explicit downcast::<Elapsed> →
    TransparentInputNotFound
  ────────────────────────────────────────
  transaction.rs:line: 647
  Query: state.oneshot(BestChainNextMedianTimePast)
  How its error reaches TransactionError: explicit map_err(... 
    ValidateMempoolLockTimeError)

  Only one of those (line 721, AwaitUtxo) is semantically a "transparent input
  not found." The fix in error.rs:317 is correct for that case — that's the real
   bug. But:

  The over-broad case: CheckBestChainTipNullifiersAndAnchors

  It rides the same Timeout<State> (same 6-min budget) and surfaces via the same
   From<BoxError> path. If that state call times out — e.g. state is overloaded
  near tip — the new arm will report TransparentInputNotFound, even though the
  transaction's transparent inputs were never the problem. It's a
  nullifier/anchor consistency check.

  Practical impact is mild (misbehavior score is 0 for both, tx just gets
  rejected/retried), but the logs and any future logic that branches on
  TransparentInputNotFound would be misled.

  Tower timeouts that don't flow through this conversion

  - Timeout::new(tx_verifier, …) in zebrad/.../mempool.rs:366 and
  Timeout::new(block_verifier, …) in inbound.rs:278 / sync.rs:496 wrap the
  verifier as a whole. Their Elapsed is observed by the caller
  (mempool/inbound/sync), not produced inside the verifier, so
  TransactionError::from(BoxError) isn't on that path.
  zebrad/src/application.rs:377 already has an is::<Elapsed> check on the outer
  side.

  Suggested narrower fix

  Mirror the AwaitOutput pattern (transaction.rs:745–750) at the AwaitUtxo call
  site (transaction.rs:721), instead of in the global From<BoxError>:

  let utxo = match query.await {
      Ok(zebra_state::Response::Utxo(utxo)) => utxo,
      Ok(_) => unreachable!("AwaitUtxo always responds with Utxo"),
      Err(err) => return match err.downcast::<tower::timeout::error::Elapsed>()
  {
          Ok(_) => Err(TransactionError::TransparentInputNotFound),
          Err(err) => Err(err.into()),
      },
  };

  That fixes the actual stall loop without silently relabelling unrelated state
  timeouts. If you do prefer the blanket conversion, at minimum it should map to
   a distinct variant (e.g. a new StateTimeout) rather than to
  TransparentInputNotFound, so callers and logs can tell the two failure modes
  apart.
```

File: zebra-consensus/src/error.rs [hunk 1]
```diff
@@ -314,6 +314,12 @@ impl From<BoxError> for TransactionError {
             Err(e) => err = e,
         }
 
+        // tower::timeout::error::Elapsed from Timeout<State> or Timeout<Mempool>
+        match err.downcast::<tower::timeout::error::Elapsed>() {
+            Ok(_) => return TransactionError::TransparentInputNotFound,
```
Comment (gustavovalverde) [inline zebra-consensus/src/error.rs]: This conversion is wider than the failure mode described in the PR. `From<BoxError> for TransactionError` is also used by other state-service checks, for example `CheckBestChainTipNullifiersAndAnchors` through `async_checks.check().await?` in `transaction.rs`. If that state calls times out, this branch reports `TransparentInputNotFound` even though no transparent input lookup failed.

I think you should move the `Elapsed -> TransparentInputNotFound` mapping to the `AwaitUtxo` call site, mirroring the existing `AwaitOutput` handling, or use a distinct timeout variant.

File: zebra-consensus/src/error.rs [hunk 2]
```diff
@@ -474,3 +480,16 @@ impl BlockError {
...
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn elapsed_timeout_converts_to_transparent_input_not_found() {
+        let elapsed: BoxError = Box::new(tower::timeout::error::Elapsed::new());
+        let tx_error = TransactionError::from(elapsed);
+
+        assert_eq!(tx_error, TransactionError::TransparentInputNotFound);
```
Comment (gustavovalverde) [inline zebra-consensus/src/error.rs]: This test locks in the broad conversion instead of the bug's call-site behavior. It would still pass for unrelated state timeouts that should not become `TransparentInputNotFound`. Please cover the `AwaitUtxo` timeout path directly and add a negative control test for a non-UTXO state timeout if this remains near the shared conversion boundary.

**Review by gustavovalverde — CHANGES_REQUESTED** [review]

**Comment by oxarbitrage** [comment]
Updated based on @conradoplg and @gustavovalverde's feedback.

**What changed:**
- Removed the broad `Elapsed` downcast arm from `From<BoxError> for TransactionError` in `error.rs`
- Added precise `Elapsed` handling at the `AwaitUtxo` call site in `transaction.rs`, mirroring the existing `AwaitOutput` pattern

**Why this is better:**
The previous approach would mislabel any timeout through `From<BoxError>` as `TransparentInputNotFound`, including `CheckBestChainTipNullifiersAndAnchors` timeouts, which are nullifier/anchor checks — not UTXO lookups. The new approach only maps `Elapsed` to `TransparentInputNotFound` at the two call sites where that mapping is semantically correct.

The two patterns are now identical:
```rust
Err(err) => return match err.downcast::<Elapsed>() {
    Ok(_) => Err(TransactionError::TransparentInputNotFound),
    Err(err) => Err(err.into()),
},
```

File: zebra-consensus/src/transaction.rs [hunk 3]
```diff
@@ -719,10 +719,15 @@ where
...
-                    if let zebra_state::Response::Utxo(utxo) = query.await? {
-                        utxo
-                    } else {
-                        unreachable!("AwaitUtxo always responds with Utxo")
+                    match query.await {
+                        Ok(zebra_state::Response::Utxo(utxo)) => utxo,
+                        Ok(_) => unreachable!("AwaitUtxo always responds with Utxo"),
+                        Err(err) => {
+                            return match err.downcast::<Elapsed>() {
+                                Ok(_) => Err(TransactionError::TransparentInputNotFound),
```
Comment (gustavovalverde) [inline zebra-consensus/src/transaction.rs:727]: Worth considering a rename to `TransparentInputUnavailable` in a follow-up, since both this path and the existing one at line 746 collapse "timed out" into this variant.

File: zebra-consensus/src/error.rs [hunk 4]
```diff
@@ -474,3 +474,8 @@ impl BlockError {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (gustavovalverde) [inline zebra-consensus/src/error.rs]: This compiles to nothing. The PR body promises an `elapsed_timeout_converts_to_transparent_input_not_found` test here.

**Review by gustavovalverde — CHANGES_REQUESTED** [review]
1. Title and body describe modifying `From<BoxError> for TransactionError` and adding a unit test `elapsed_timeout_converts_to_transparent_input_not_found`, but neither appears in the diff.

2. Issue #10629 itself recommends putting the `Elapsed` arm in `From<BoxError>` so the existing `downcast::<Elapsed>()` block at `transaction.rs:746` (mempool path) becomes redundant. This PR instead duplicates that pattern at a second site. Centralizing the conversion is cleaner and forecloses the same gap that would appear at any future third-timeout site.

3. User-visible behavior, so it needs a `CHANGELOG.md [Unreleased]` entry.

File: zebra-consensus/src/error.rs (hunk 4 above)
Comment (oxarbitrage) [reply zebra-consensus/src/error.rs]: Sorry about that, it was leftover, Fixed in https://github.com/ZcashFoundation/zebra/pull/10630/commits/475edb6fb2c49ff3ebc8f58cf552cc5be67415f0

**Review by conradoplg — CHANGES_REQUESTED** [review]
I think this would need further adjustments because somewhere in the code (can't remember the exact location right now, sorry) it checks for specific errors to determine if the sync should be restarted and the new error isn't in it:


```
2026-05-26T14:47:13.284639Z ERROR sync:try_to_sync:try_to_sync_once: zebrad::components::sync: a BlockDownloadVerifyError that should have been filtered out was detected, which possibly indicates a programming error in the downcast inside zebrad::components::sync::downloads::Downloads::download_and_verify e=Invalid { error: Block { source: Transaction(TransparentInputNotFound) }, height: Height(3332673), hash: block::Hash("0000000000f1b8120d5680df25c295d56cd189c8b6b81588d543ac2b7bdd21d1"), advertiser_addr: Some(v4redacted:8233) }
2026-05-26T14:47:13.284661Z  WARN sync:try_to_sync:try_to_sync_once: zebrad::components::sync: error downloading and verifying block e=Invalid { error: Block { source: Transaction(TransparentInputNotFound) }, height: Height(3332673), hash: block::Hash("0000000000f1b8120d5680df25c295d56cd189c8b6b81588d543ac2b7bdd21d1"), advertiser_addr: Some(v4redacted:8233) }
```

**Comment by oxarbitrage** [comment]
Updated based on review feedback:

- Removed the empty `#[cfg(test)] mod tests` block from `error.rs` (leftover from the first approach)
- Added a unit test `block_request_with_await_utxo_timeout_returns_transparent_input_not_found` that mocks `AwaitUtxo` returning `Elapsed` and asserts it maps to `TransparentInputNotFound`, not `InternalDowncastError`
- Added CHANGELOG entry under `[Unreleased] > Fixed`
- Updated PR title and description to reflect the call-site approach

**Comment by oxarbitrage** [comment]
@conradoplg Good catch — you were right that `TransparentInputNotFound` was hitting the `_` catch-all in `should_restart_sync`, triggering the "should have been filtered out" error and restarting sync anyway.

Fixed in 5967bd904:

- Added `is_transparent_input_not_found()` to `VerifyBlockError` and `RouterError`
- Added an explicit arm in `should_restart_sync` that matches `BlockDownloadVerifyError::Invalid` when the inner error is `TransparentInputNotFound`, logs at debug level, and returns `false` (no sync restart)

So now the full path works: `AwaitUtxo` timeout → `TransparentInputNotFound` → sync continues instead of restarting.

**Review by conradoplg — APPROVED** [review]

File: zebra-consensus/src/transaction.rs (hunk 3 above)
Comment (conradoplg) [reply zebra-consensus/src/transaction.rs:727]: Agree this can be improved a bit, but let's get this merged

**Comment by oxarbitrage** [comment]
Closing in favor of #10442, which addresses the sync stall at the network/peer routing level — height-aware peer selection prevents stalled peers from ever receiving block requests, making the AwaitUtxo timeout path unreachable in practice during normal sync. The underlying timeout handling can be revisited as a hardening fix if needed post-merge.
